### PR TITLE
[Expert Filter] : Support StaticVarCompensator equipment type

### PR DESCRIPTION
**Pull request by thangqp**
Related PRs:

https://github.com/gridsuite/filter/pull/17
https://github.com/gridsuite/filter-server/pull/117
https://github.com/gridsuite/gridexplore-app/pull/439

Demo

svar_filter_expert

### CODE REVIEW

File: src/components/filter/expert/expert-filter-utils.ts [hunk 1]
```diff
@@ -407,6 +519,22 @@ export const queryValidator: QueryValidator = (query) => {
                 valid: false,
                 reasons: [RULES.EMPTY_RULE],
             };
+        } else if (rule.id && dataType === DataType.COMBINATOR) {
+            // based on FIELDS_OPTIONS configuration and group rule value, validate for each child rule in a group rule
+            const childFields = Object.keys(
```
Comment (Mathieu-Deharbe) [inline src/components/filter/expert/expert-filter-utils.ts:524]:
```suggestion
            const childrenFields = Object.keys(
```

File: src/components/inputs/react-query-builder/group-rule-editor/rule-value-editor.tsx [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+                </Select>
+            </Grid>
+            <Grid
+                container
+                item
+                xs={5.5}
+                paddingLeft={1}
+                direction={'column'}
+                justifyContent={'flex-end'}
+                alignItems={'baseline'}
```
Comment (Mathieu-Deharbe) [inline src/components/inputs/react-query-builder/group-rule-editor/rule-value-editor.tsx:97]: Used 3 times. Might be a style ?

File: src/components/translations/filter-expert-en.ts [hunk 3]
```diff
@@ -70,6 +70,29 @@ const filter_expert_en = {
...
+    fixQAtNominalV: 'Fixed part of Q at nominal voltage ',
+    'maxSusceptance.svar': 'Susceptance max',
+    minSusceptance: 'Susceptance min',
+    svarRegulationMode: 'Regulation mode',
+    'svarRegulationMode.off': 'Off',
+    'svarRegulationMode.voltage': 'Voltage regulation',
+    'svarRegulationMode.reactivePower': 'Reactive power regulation',
+    voltageSetPoint: 'Voltage set point',
+    reactivePowerSetPoint: 'Reactive power set point',
+    regulatingTerminal: 'Regulated terminal',
```
[the remark this answers is not in the document]
Comment (thangqp) [reply src/components/translations/filter-expert-en.ts:84]: Yes, it is "Regulating terminal", the svar is regulated

File: src/components/filter/expert/expert-filter.type.ts [hunk 4]
```diff
@@ -119,5 +142,22 @@ export interface RuleTypeExport {
...
+    children?: { [field: string]: FullField };
+}
+
+// typing group rule value
+export interface GroupRuleValue {
+    operator: string;
+    value: string;
+}
+export interface GroupRule {
+    [field: string]: GroupRuleValue;
```
[the remark this answers is not in the document]
Comment (thangqp) [reply src/components/filter/expert/expert-filter.type.ts:162]: I will rename GroupRuleValue => CompositeRule, GroupRule => CompositeGroup and GroupRuleField => CompositeFiled

File: src/components/filter/expert/expert-filter-constants.ts [hunk 5]
```diff
@@ -53,6 +53,10 @@ export const EXPERT_FILTER_EQUIPMENTS = {
         id: 'SHUNT_COMPENSATOR',
         label: 'ShuntCompensators',
     },
+    STATIC_VAR_COMPENSATOR: {
+        id: 'STATIC_VAR_COMPENSATOR',
+        label: 'StaticVarCompensators',
+    },
```
[the remark this answers is not in the document]
Comment (thangqp) [reply src/components/filter/expert/expert-filter-constants.ts:59]: DONE in https://github.com/gridsuite/commons-ui/pull/452/commits/f8a71a4e4be081d578720514b66b295e3d65ef93

File: src/components/translations/filter-expert-fr.ts [hunk 6]
```diff
@@ -70,6 +70,29 @@ const filter_expert_fr = {
...
+    minSusceptance: 'Susceptance min',
+    svarRegulationMode: 'Mode de réglage',
+    'svarRegulationMode.off': 'Arrêt',
+    'svarRegulationMode.voltage': 'Réglage de tension',
+    'svarRegulationMode.reactivePower': 'Réglage du réactif',
+    voltageSetPoint: 'Tension',
+    reactivePowerSetPoint: 'Puissance réactif',
+    regulatingTerminal: 'Terminal distant réglé',
+    regulatingTerminalVLId: 'Poste ID',
+    regulatingTerminalConnectableId: 'Ouvrage ID',
```
[the remark this answers is not in the document]
Comment (thangqp) [reply src/components/translations/filter-expert-fr.ts]: DONE in https://github.com/gridsuite/commons-ui/pull/452/commits/f8a71a4e4be081d578720514b66b295e3d65ef93

File: src/components/translations/filter-expert-en.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (thangqp) [reply src/components/translations/filter-expert-en.ts]: DONE in https://github.com/gridsuite/commons-ui/pull/452/commits/f8a71a4e4be081d578720514b66b295e3d65ef93

File: src/components/filter/expert/expert-filter-utils.ts [hunk 7]
```diff
@@ -37,32 +40,71 @@ import {
...
+interface TreeNode {
+    [key: string]: any;
+    children?: Tree;
+}
+
+interface Tree {
+    [key: string]: TreeNode;
+}
+
+const search = (tree: Tree, value: any, key: string) => {
```
[the remark this answers is not in the document]
Comment (thangqp) [reply src/components/filter/expert/expert-filter-utils.ts]: DONE in https://github.com/gridsuite/commons-ui/pull/452/commits/f8a71a4e4be081d578720514b66b295e3d65ef93

File: src/components/filter/expert/expert-filter.type.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (thangqp) [reply src/components/filter/expert/expert-filter.type.ts]: DONE in https://github.com/gridsuite/commons-ui/pull/452/commits/f8a71a4e4be081d578720514b66b295e3d65ef93

File: src/components/filter/expert/expert-filter-utils.ts (hunk 1 above)
Comment (thangqp) [reply src/components/filter/expert/expert-filter-utils.ts]: DONE in https://github.com/gridsuite/commons-ui/pull/452/commits/f8a71a4e4be081d578720514b66b295e3d65ef93

File: src/components/inputs/react-query-builder/group-rule-editor/rule-value-editor.tsx (hunk 2 above)
Comment (thangqp) [reply src/components/inputs/react-query-builder/group-rule-editor/rule-value-editor.tsx]: DONE in https://github.com/gridsuite/commons-ui/pull/452/commits/707406d3e41b369c329831e0523a672508d466d6

**Review by Mathieu-Deharbe — APPROVED** [review]

File: src/components/translations/filter-expert-fr.ts [hunk 8]
```diff
@@ -70,6 +70,29 @@ const filter_expert_fr = {
     voltageLevelProperty: 'Propriété poste',
     voltageLevelProperty1: 'Propriété poste 1',
     voltageLevelProperty2: 'Propriété poste 2',
+    'maxQAtNominalV.svar': 'Q max à tension nominale',
+    minQAtNominalV: 'Q min à tension nominale',
+    fixQAtNominalV: 'Q fixe à tension nominale',
+    'maxSusceptance.svar': 'Susceptance max',
+    minSusceptance: 'Susceptance min',
+    svarRegulationMode: 'Mode de réglage',
```
[the remark this answers is not in the document]
Comment (thangqp) [reply src/components/translations/filter-expert-fr.ts:78]: There are ratioRegulationMode and phaseRegulationMode at line 58-59, that why I use svar as prefix for  Static Var Compensator RegulationMode

File: src/components/translations/filter-expert-fr.ts [hunk 9]
```diff
@@ -70,6 +70,29 @@ const filter_expert_fr = {
...
+    'svarRegulationMode.off': 'Arrêt',
```
[the remark this answers is not in the document]
Comment (thangqp) [reply src/components/translations/filter-expert-fr.ts:79]: Idem, see my reply above

File: src/components/translations/filter-expert-en.ts [hunk 10]
```diff
@@ -70,6 +70,29 @@ const filter_expert_en = {
     voltageLevelProperty: 'Voltage level property',
     voltageLevelProperty1: 'Voltage level property 1',
     voltageLevelProperty2: 'Voltage level property 2',
+    'maxQAtNominalV.svar': 'Q max at nominal voltage',
```
[the remark this answers is not in the document]
Comment (thangqp) [reply src/components/translations/filter-expert-en.ts:73]: Two translation keys ended with suffix '.svar' in order to customize label without unit (demanded by Christelle)

[the remark this answers is not in the document]
Comment (thangqp) [reply src/components/translations/filter-expert-en.ts]: DONE! Unit are added to some label

File: src/components/translations/filter-expert-fr.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (thangqp) [reply src/components/translations/filter-expert-fr.ts]: DONE! Finally, I correct these keys as your suggestion.. see https://github.com/gridsuite/commons-ui/pull/452/commits/6ae6cfcfa3b89bc4ac6bcd202044cf480eeaf9c7

File: src/components/translations/filter-expert-fr.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (thangqp) [reply src/components/translations/filter-expert-fr.ts]: DONE! corrected in https://github.com/gridsuite/commons-ui/pull/452/commits/6ae6cfcfa3b89bc4ac6bcd202044cf480eeaf9c7

**Review by etiennehomer — APPROVED** [review]

**Review by Mathieu-Deharbe — APPROVED** [review]

**Review by Mathieu-Deharbe — APPROVED** [review]

**Pull request merged**